core/shared/src/main/scala/scalaz/zio/stm/STM.scala

```diff
@@ -0,0 +1,584 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scalaz.zio.stm
+
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicLong, AtomicReference }
+
+import scalaz.zio.{ IO, UIO }
+
+import scala.collection.mutable.{ Map => MutableMap }
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * `STM[E, A]` represents an effect that can be performed transactionally,
+ * resulting in a failure `E` or a value `A`.
+ *
+ * {{{
+ * def transfer(receiver: TRef[Int],
+ *              sender: TRef[Int], much: Int): UIO[Int] =
+ *   STM.atomically {
+ *     for {
+ *       balance <- sender.get
+ *       _       <- STM.check(balance >= much)
+ *       _       <- receiver.update(_ + much)
+ *       _       <- sender.update(_ - much)
+ *       newAmnt <- receiver.get
+ *     } yield newAmnt
+ *   }
+ *
+ *   val action: UIO[Int] =
+ *     for {
+ *       t <- STM.atomically(TRef.make(0).zip(TRef.make(20000)))
+ *       (receiver, sender) = t
+ *       balance <- transfer(receiver, sender, 1000)
+ *     } yield balance
+ * }}}
+ */
+final class STM[+E, +A] private[stm] (
+  val exec: STM.internal.Journal => STM.internal.TRez[E, A]
+) extends AnyVal { self =>
+  import STM.internal.{ resetJournal, TRez }
+
+  /**
+   * Sequentially zips this value with the specified one.
+   */
+  final def <*>[E1 >: E, B](that: => STM[E1, B]): STM[E1, (A, B)] =
```

Comment:
Hmm, personally I'm pretty happy with `zip` for this operation without a symbolic alias. Moreover, `<*>` is not what I'd expect for it :-)

Reply:
- `<*` — zip, take left
- `*>` — zip, take right
- `<*>` — zip, take both

The `*` is product, but is too overloaded, `<` `>` are used for applicative operators, hence `<*>`. It's also worth noting the canonical applicative operator in Haskell is `ap` which has the alias `<*>`, but in Scala (without curried functions), it's actually zip, which by all rights should use the same operator.

Reply:
Ok, I'm sold on the consistency with `*>` and `<*`. I think `ZIO` is missing that currently. Would also be nice to have `<&>` etc. for the `Par` versions.

Reply:
That would be great. I'll add tickets, both to add `<*>` as an alias for `zip`, and add `<&>` as an alias for `zipPar`, pervasively over all the data types that support them.

Reply:
`>*< :: f a -> f b -> f (a, b)`
https://hackage.haskell.org/package/invertible-0.2.0.5/docs/Control-Invertible-Monoidal.html#v:-62--42--60-